### PR TITLE
RDKB-57495: XDNS DM implementation

### PR DESCRIPTION
**Pull request by KavyaShivalingaiah**
Reason for change: Implement DM to notify DNS changes to XDNS

Test Procedure:
Updated in Jira.

Risks: none
Priority: P1

### CODE REVIEW

File: source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.h [hunk 1]
```diff
@@ -61,6 +61,7 @@ struct IFACE_INFO
     CHAR    ActiveStatus[BUFLEN_64];
     CHAR    CurrentActive[BUFLEN_64];
     CHAR    CurrentStandby[BUFLEN_64];
+    CHAR    CurrentActiveDNS[BUFLEN_256];
```
Comment (S-Parthiban-Selvaraj) [inline source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.h:64]: IFACE_INFO linked list was initially added to handle multiple interfaces AvailableStatus in a sorted order. 
CurrentActiveDNS will be needed only for an active interface. Should we require it in the linked list?

File: source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c [hunk 2]
```diff
@@ -1933,6 +1984,10 @@ ANSC_STATUS Update_Interface_Status()
     {
         WanMgr_Rbus_String_EventPublish_OnValueChange(WANMGR_CONFIG_WAN_INTERFACEACTIVESTATUS, prevInterfaceActiveStatus, InterfaceActiveStatus);
     }
+    if(publishCurrentActiveDNS == TRUE)
+    {
+        WanMgr_Rbus_String_EventPublish_OnValueChange(WANMGR_CONFIG_WAN_CURRENTACTIVEDNS, prevCurrentActiveDNS, CurrentActiveDNS);
```
Comment (S-Parthiban-Selvaraj) [inline source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c:1989]: Is the OLD value required in the publish?

**Review by S-Parthiban-Selvaraj — CHANGES_REQUESTED** [review]

File: source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c [hunk 3]
```diff
@@ -1721,7 +1722,47 @@ void SortedInsert( struct IFACE_INFO** head_ref,  struct IFACE_INFO *new_node)
...
+{
+    FILE *fp = NULL;
+    char buf[64] = {0};
+    char* token = NULL;
 
+    memset(CurrentActiveDNS,0,sizeof(CurrentActiveDNS));
+    if((fp = fopen(RESOLV_CONF_FILE, "r")) == NULL)
+    {
+        CcspTraceError(("%s %d - Open %s error!\n", __FUNCTION__, __LINE__, RESOLV_CONF_FILE));
+        return RETURN_ERR;
```
Comment (S-Parthiban-Selvaraj) [inline source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c:1735]: The function has void return type.

File: source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c [hunk 4]
```diff
@@ -1874,6 +1919,16 @@ ANSC_STATUS Update_Interface_Status()
             publishActiveStatus = TRUE;
 #endif
         }
+        if(strcmp(pWanDmlData->CurrentActiveDNS,CurrentActiveDNS) != 0)
```
Comment (S-Parthiban-Selvaraj) [inline source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c:1922]: Check for empty CurrentActiveDNS. Update_Current_ActiveDNSmay fail to open the dns conf and return empty string

File: source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c [hunk 5]
```diff
@@ -1874,6 +1919,16 @@ ANSC_STATUS Update_Interface_Status()
...
+        {
+	    CcspTraceInfo(("%s %d CurrentActiveDNS- [%s] [%s]\n",__FUNCTION__,__LINE__,pWanDmlData->CurrentActiveDNS,CurrentActiveDNS));
+	    strncpy(prevCurrentActiveDNS,pWanDmlData->CurrentActiveDNS,sizeof(prevCurrentActiveDNS)-1);
+            memset(pWanDmlData->CurrentActiveDNS,0, sizeof(pWanDmlData->CurrentActiveDNS));
+            strncpy(pWanDmlData->CurrentActiveDNS,CurrentActiveDNS, sizeof(pWanDmlData->CurrentActiveDNS) - 1);
+#ifdef RBUS_BUILD_FLAG_ENABLE
+            publishCurrentActiveDNS = TRUE;
```
Comment (S-Parthiban-Selvaraj) [inline source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c:1929]: Directly publish here. 
publishCurrentActiveDNS is not necessary.

File: source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c [hunk 6]
```diff
@@ -1721,7 +1722,47 @@ void SortedInsert( struct IFACE_INFO** head_ref,  struct IFACE_INFO *new_node)
...
+    if((fp = fopen(RESOLV_CONF_FILE, "r")) == NULL)
+    {
+        CcspTraceError(("%s %d - Open %s error!\n", __FUNCTION__, __LINE__, RESOLV_CONF_FILE));
+        return RETURN_ERR;
+    }
+
+    while((fgets(buf, sizeof(buf), fp)) != NULL)
+    {
+        token = strtok(buf, " ");
+	if(!strcmp(token,"nameserver"))
```
Comment (S-Parthiban-Selvaraj) [inline source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c:1741]: Please fix the indentation

File: source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c [hunk 7]
```diff
@@ -1721,7 +1722,47 @@ void SortedInsert( struct IFACE_INFO** head_ref,  struct IFACE_INFO *new_node)
...
+                }
+		if(token[strlen(token)-1] == '\n')
+		{
+                    token[strlen(token)-1] = '\0';
+		}
+                strcat(CurrentActiveDNS,token);
+            }
+	}
+        memset(buf, 0 , sizeof(buf));
+        memset(token,0,sizeof(token));
```
Comment (S-Parthiban-Selvaraj) [inline source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c:1758]: the token is a char pointer. memset is necessary?

**Review by S-Parthiban-Selvaraj — CHANGES_REQUESTED** [review]

**Review by S-Parthiban-Selvaraj — APPROVED** [review]

**Review by guto86 — APPROVED** [review]

**Pull request merged**